Add minimum projects rule to Budgets

agustibr opened the pull request:
#### :tophat: What? Why?

Add an option for admins so they can configure a Budget component rule to a minimum number of projects that participants must select to vote.

The participant has to select at least the minimum number of projects to vote.

#### :pushpin: Related Issues
- Fixes #5754

#### :clipboard: Subtasks
- [x] Add `CHANGELOG` entry
- [x] Add tests


### :camera: Screenshots (optional)

<img width="976" alt="Screenshot 2020-03-16 at 19 04 42" src="https://user-images.githubusercontent.com/210216/76787531-8539f480-67b9-11ea-8c86-a8ff92a01ce7.png">  

<img width="961" alt="Screenshot 2020-03-16 at 19 06 57" src="https://user-images.githubusercontent.com/210216/76787548-9256e380-67b9-11ea-9520-1498b1ac6e7a.png">

**Admin**

<img width="706" alt="Screenshot 2020-03-16 at 19 20 50" src="https://user-images.githubusercontent.com/210216/76788554-72c0ba80-67bb-11ea-8683-a8ae756fb12b.png">

<img width="708" alt="Screenshot 2020-03-17 at 12 56 03" src="https://user-images.githubusercontent.com/210216/76855942-810de580-6852-11ea-97ed-cc3bc5226d72.png">

<img width="705" alt="Screenshot 2020-03-17 at 12 56 45" src="https://user-images.githubusercontent.com/210216/76855956-88cd8a00-6852-11ea-8dbf-259d0922f8f7.png">

mrcasals commented on decidim-admin/app/forms/decidim/admin/component_form.rb:
```diff
@@ -90,6 +90,8 @@ def budget_voting_rule_enabled_setting
 
       # - the value must be a valid number
       def budget_voting_rule_value_setting
+        return unless manifest&.name == :budgets
```
I think we should try to separate this logic and extract it from here, so that we don't link the admin component with the budgets one. 

It would require a little refactor, but what about making the component manifest define an optional class, something like `MyComponent::ComponentSettingsForm` that receives the settings and validates them? The `Admin::ComponentForm` would find this class via the component manifest, call it if present and check if it's valid. If it's not present, then we go ahead. If it's present and valid, then go ahead. If it's present and not valid, we return the errors. This would open the possibility for other components to implement extra logic on settings.

What do you think? Would that work for you?

agustibr replied:
Yes, I like the idea, I'll refactor it. Thanks @mrcasals 😃

leio10 commented on decidim-budgets/app/commands/decidim/budgets/add_line_item.rb:
```diff
@@ -10,7 +10,7 @@ class AddLineItem < Rectify::Command
       # current_user - The current user logged in
       def initialize(current_order, project, current_user)
-        @order = current_order
+        @order = current_order.persisted? ? current_order : nil
```
Sorry, but I can't find out why this change is needed for this feature. Is this fixing a new bug found?

agustibr replied:
Yes found a bug and needed it 😉.
Bug: When €0 assigned it didn't show the threshold nor the minimum amount the participant required to vote. [screenshot €0 assigned](https://user-images.githubusercontent.com/210216/77104851-f7f6da00-6a1c-11ea-833f-79330837173b.png) and [screenshot with € assigned](https://user-images.githubusercontent.com/210216/77104995-34c2d100-6a1d-11ea-805d-edb2168780c7.png).

Before (this PR) the `current_order` could be nil and was created in the commands private method `:order` when the first project was added.

The order is needed before to know the rules to show to the participant, so now the `current_order` always returns an order object.

```rb
   @current_order ||= Order.find_by || Order.new
```

So this way the command works, and fixes the issue. 😁

https://github.com/decidim/decidim/blob/fbabd9e3513d38617325d901ba66695443a02514/decidim-budgets/app/controllers/concerns/decidim/budgets/needs_current_order.rb#L17-L19

leio10 replied:
Sorry, I didn't see your question before. I think that is solved, but I can think in some improvements on it:
* You can use `find_or_initialize_by` instead of `find_by || new`, in this case I think that it would do the exactly the same that you are doing
* I wouldn't add the `persisted?` related logic to the command. I'd add a `persisted_current_order` getter to the controller and I'd use it when calling the command.

Does this help you? 😬

agustibr replied:
thanks @leio10 , modified as you suggested. 
Yes, Certainly `find_or_initialize_by` is for this use cases.

leio10 commented on decidim-budgets/app/helpers/decidim/budgets/projects_helper.rb:
```diff
@@ -16,6 +16,28 @@ def current_order_budget_percent
+      def budget_confirm_disabled_attr
+        return if current_order_can_be_checked_out?
+
+        %( disabled= "disabled" )
```
I didn't remember the `%(` meaning and I was thinking that you were trying to use the setter of a `disabled` property. 😅 
Maybe we can remove the extra space: `%( disabled="disabled" )`. BTW, wouldn't `disabled` be enough?

agustibr replied:
Yes I'll remove the extra space.
```
%( disabled="disabled" ) == "disabled=\"disabled\""
```
I think this way is more compatible, same as `checked="checked"` for checkboxes. I'll find out.

tramuntanal replied:
```suggestion
        %( disabled="disabled" )
```

leio10 commented on decidim-admin/app/assets/javascripts/decidim/admin/form.js.es6:
```diff
@@ -1,3 +1,16 @@
+
+  budgetRuleToggler.run();
+})(window);
+
```
Nice hack! 😎

agustibr commented:
@mrcasals I've refactored the compnent settings so the logic its on its module, detaching it from core/admin 😃
You can check it on the commit:

- [move to Budgets module extra logic of settings validation](https://github.com/decidim/decidim/pull/5865/commits/19252ec384a04843091fc4798861c4a5233eb460)

As Proposals had some amendments and participatory_texts logic, I've moved it to its module:

- [move to Proposals extra logic of settings validations](https://github.com/decidim/decidim/pull/5865/commits/deb7ef90f0edb5104fbcded5e407d6fdb8da40de)

agustibr commented:
@decidim/core please review when you can, this has been approved by @decidim/product.

tramuntanal commented on decidim-budgets/app/forms/decidim/budgets/admin/component_form.rb:
```diff
@@ -0,0 +1,44 @@
+module Decidim
+  module Budgets
+    module Admin
+      # A form object for Budgets used to attach a component to a participatory process from the
```
```suggestion
      # A form object for the budgets component. Used to attach the component to a participatory process from the
```

tramuntanal commented on decidim-budgets/app/commands/decidim/budgets/add_line_item.rb:
```diff
@@ -10,7 +10,7 @@ class AddLineItem < Rectify::Command
       # current_user - The current user logged in
       def initialize(current_order, project, current_user)
-        @order = current_order
+        @order = current_order&.persisted? ? current_order : nil
```
if current_order is passed but not persisted it will create a new Order. Is it correct?

agustibr replied:
Yes, the _why_ is explained in an answer to Leo => https://github.com/decidim/decidim/pull/5865#discussion_r395261728

agustibr replied:
@tramuntanal refactored as suggested by @leio10 => https://github.com/decidim/decidim/pull/5865#discussion_r401620195

tramuntanal commented on decidim-admin/app/forms/decidim/admin/component_form.rb:
```diff
@@ -54,26 +51,6 @@ def settings_errors_empty?
         validations.all?
       end
-
-      # Validates setting `participatory_texts_enabled` is not changed when there are proposals for the component.
```
ouyeah!

agustibr commented:
@tramuntanal, I've applied suggested changes, it should be ready to go 😁
